src/NHibernate/Engine/Query/QueryPlanCache.cs

```diff
@@ -200,46 +199,24 @@ protected HQLQueryPlanKey(System.Type queryTypeDiscriminator, string query, bool
 				unchecked
 				{
 					var hash = query.GetHashCode();
-					hash = 29*hash + (shallow ? 1 : 0);
-					hash = 29*hash + CollectionHelper.GetHashCode(filterNames);
-					hash = 29*hash + queryTypeDiscriminator.GetHashCode();
+					hash = 29 * hash + (shallow ? 1 : 0);
+					hash = 29 * hash + CollectionHelper.GetHashCode(filterNames);
+					hash = 29 * hash + queryTypeDiscriminator.GetHashCode();
 					hashCode = hash;
 				}
 			}
 
-			public override bool Equals(object obj)
+			public override bool Equals(QueryPlanKey other)
```

Comment:
Overrides the base one for first calling it and ruling out simple inequality cases (not same hashcode or not same query) then if deemed equal by base implementation, checking the additional constraints of the subclass.